src/libraries/System.Net.Http/tests/FunctionalTests/MetricsTest.cs

```diff
@@ -617,24 +617,30 @@ public async Task RequestDuration_EnrichmentHandler_ContentLengthError_Recorded(
         [Fact]
         public async Task FailedRequests_ConnectionClosedWhileReceivingHeaders_Recorded()
         {
-            TimeSpan timeout = TimeSpan.FromSeconds(30);
             await LoopbackServerFactory.CreateClientAndServerAsync(async uri =>
             {
                 using HttpMessageInvoker client = CreateHttpMessageInvoker();
                 using InstrumentRecorder<long> recorder = SetupInstrumentRecorder<long>(InstrumentNames.FailedRequests);
                 using HttpRequestMessage request = new(HttpMethod.Get, uri) { Version = UseVersion };
 
-                await Assert.ThrowsAsync<HttpRequestException>(async () =>
+                Exception ex = await Assert.ThrowsAnyAsync<Exception>(async () =>
                 {
-                    using HttpResponseMessage response = await SendAsync(client, request);
-                }).WaitAsync(timeout);
+                    // Getting a cancellation is also good if we are unable to detect the peer shutdown.
+                    using CancellationTokenSource cts = new CancellationTokenSource(1000);
+                    using HttpResponseMessage response = await SendAsync(client, request, cts.Token);
+                });
+                Assert.True(ex is HttpRequestException or TaskCanceledException);
 
                 Measurement<long> m = recorder.GetMeasurements().Single();
                 VerifyFailedRequests(m, 1, uri, null, null);
             }, async server =>
             {
-                var connection = (LoopbackServer.Connection)await server.EstablishGenericConnectionAsync().WaitAsync(timeout);
-                connection.Socket.Close();
+                try
+                {
+                    var connection = (LoopbackServer.Connection)await server.EstablishGenericConnectionAsync().WaitAsync(20_000);
```

Comment:
same as above.

Reply:
This should be actually cancelled when the client request completes, changed the test.